core/preconditioner/jacobi.cpp

```diff
@@ -231,11 +231,23 @@ void Jacobi<ValueType, IndexType>::generate(const LinOp *system_matrix,
     using csr_type = matrix::Csr<ValueType, IndexType>;
     const auto exec = this->get_executor();
     if (parameters_.max_block_size == 1) {
-        auto diag = as<DiagonalExtractable<ValueType>>(system_matrix)
-                        ->extract_diagonal();
-        auto temp = gko::Array<ValueType>::view(
-            exec, system_matrix->get_size()[0], diag->get_values());
-        this->blocks_ = temp;
+        auto diag = share(as<DiagonalLinOpExtractable>(system_matrix)
+                              ->extract_diagonal_linop());
```

Comment:
Is there some reason we need the `extract_diagonal_linop` and `extract_diagonal` will not work ?

Reply:
extract_diagonal require the valuetype, but extract_diagonal_linop does not. we can get the diag first and then check which precision to work.
I am thinking do the dynamic_cast to check the matrix type and use corresponding extract_diag.

Reply:
An alternative to DiagonalLinOpExtractable would be using
```
auto csr_mtx = convert_to_with_sorting<csr_type>(exec, system_matrix, /*skip_sorting=*/true);
```
from the non-scalar branch and extracting from that. That would also make for more consistent behavior between scalar and block Jacobi.

Reply:
I would like to avoid the converting matrix self because we need to pay more for it against converting diag

Reply:
Could we also not convert the extracted diagonal after calling the `extract_diagonal`. Currently, `Diagonal` does not have mixed precision convert methods, but we could easily add those

Reply:
Oh yes, I think I like that idea best of all suggestions yet :)

Reply:
but the issue is different, right?
you still need to dynamic_cast to the correct type first.
and it is the same impl in the current one if we use default behavior
if we use direct mapping, we need to add all matrix to support it (at least csr)

Reply:
Actually, i am a bit confused as to how the current implementation works. `extract_diagonal_linop` should return a `Diagonal` of `ValueType` right ? Because it is just calling the `extract_diagonal` of `Csr<ValueType, IndexType>`. So, the else if branch in line 243, will never be taken ?

Reply:
You could use something similar to `make_temporary_conversion`. The framework also supports other matrix types, you only need to wrap it like you see in that function. Precision conversions in Diagonal is definitely a good idea, I forgot to suggest those when we added the format.